### PR TITLE
Ignore inner class scanning in the compiler to support dynamic languages like groovy [SI-7741]

### PR DESCRIPTION
**Pull request by gerritjvv**
related to https://issues.scala-lang.org/browse/SI-7741

### CODE REVIEW

**Comment by retronym** [comment]
Thanks for tackling this. 

This pr  needs tests, more detailed comments (include some info from the ticket, commit subject should start with "SI-1234 " and more careful diffs (white space and indentation). For more info on our expectations, see CONTRIBUTING.MD in the root of this repo.

**Comment by gerritjvv** [comment]
Thanks,
I'll work on some unit tests, and also realized that I needed to start a scala-internals discussion. I'll do the needed work and make another pull request.

**Comment by retronym** [comment]
Okay, closing this one for now. Hope to hear from you soon! :)

File: src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala [hunk 1]
```diff
@@ -43,6 +43,7 @@ trait StandardScalaSettings {
   val unchecked =      BooleanSetting ("-unchecked", "Enable additional warnings where generated code depends on assumptions.")
   val uniqid =         BooleanSetting ("-uniqid", "Uniquely tag all identifiers in debugging output.")
   val usejavacp =      BooleanSetting ("-usejavacp", "Utilize the java.class.path in classpath resolution.")
+  val ignoreinnercls =  BooleanSetting ("-ignoreinnercls", "Ignore inner class scanning, this is required when integrating with dynamic languages where inner classes may not exist")
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:46]: This should be a -Y option -- "-Yignore-missing-inner-class"?

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 2]
```diff
@@ -131,21 +134,25 @@ abstract class ClassfileParser {
...
+        // this would give incorrect results (see SI-5031 in separate compilation scenario)
+        this.staticModule = if (root.isModule) root else root.companionModule
+        this.isScala      = false
+        parseHeader()
+        this.pool = newConstantPool
+        parseClass()
+      }
+    }catch{
+       case t:Throwable => {
+              if(!settings.ignoreinnercls.value) throw t
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:152]: Catching Throwable is not great practice. Why is it still necessary with the fix in `parseHeader`?

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 3]
```diff
@@ -1036,10 +1043,11 @@ abstract class ClassfileParser {
     for (entry <- innerClasses.entries) {
       // create a new class member for immediate inner classes
       if (entry.outerName == currentClass) {
-        val file = classPath.findSourceFile(entry.externalName.toString) getOrElse {
-          throw new AssertionError(entry.externalName)
-        }
-        enterClassAndModule(entry, file)
+        val file = classPath.findSourceFile(entry.externalName.toString).get
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:1046]:
```
classPath.findSourceFile(entry.externalName.toString) match {
   case Some(file) =>  enterClassAndModule(entry, file)
   case None if settings.YignoreMissingInnerClass =>  debugwarn(s"Class file for ${entry.externalName} not found")
   case _ => throw new AssertionError(entry.externalName)
}
```

File: src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala (hunk 1 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:46]: Why would it be `Y` option? If we commit to supporting this feature it should be `X` option, right?

Comment (gerritjvv) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:46]: I can make this either Y or X whats the difference between the two?
ignore-missing-inner-class sounds like a better naming for it, thanks,

just one issue with the naming '-' is not a valid char, or am I missing something? I can make is ignoreinnermissingcls,

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala (hunk 2 above)
Comment (gerritjvv) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:152]: I do a try catch Throwable and then re-throw so no information is lost, unless the settings ignoreinnercls is true. 

With the fix in parseHeader I still get (class java.util.NoSuchElementException/None.get). I wanted to do the least disruptive thing to the compiler for this implementation, any help on taking this out and putting a better fix into the parseHeader or understanding why and exception is still thrown would be great.

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala (hunk 3 above)
Comment (gerritjvv) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:1046]: thanks, I'll use the code you've provided, it looks much cleaner.

File: src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala (hunk 1 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:46]: Consider -Xlenient-classfiles for a more general option that might include future relaxations.  Or -Xrelax, -Xchill, or -Xcali-style.

There's not much guidance in ScalaSettings on camel case or whether X/Y prefix is desirable. My taste is for `overrideObjects`.  Then in code, you don't care if it's X/Y or whether the option name has embedded dashes.

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:46]: The difference between X and Y is that X is advanced but supported option and Y is experimental option that is not supported and can go away at any time without any prior deprecation.

**Pull request closed without merging**